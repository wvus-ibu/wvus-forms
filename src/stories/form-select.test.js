import React from "react";
import renderer from "react-test-renderer";
import { configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
configure({ adapter: new Adapter() });
import { mount } from "enzyme";
import sinon from "sinon";

import {
  SelectForm,
  SelectFormPreselected,
  SelectFormNoConfig
} from "./shared/select-control-form";
import {
  formMethodsUnTouched,
  formMethodsWithSuccess,
  formMethodsWithError,
  formMethodsOptional
} from "./shared/props-mocks";
import { WVUSForm, Message, validateRequired, SelectControl } from "../index";

const Form = WVUSForm(SelectForm);
const FormPreselected = WVUSForm(SelectFormPreselected);

/**
 * SNAPSHOTS
 */

test("Snapshot:  Select control renders correctly with no selection", () => {
  const tree = renderer.create(<Form />).toJSON();
  expect(tree).toMatchSnapshot();
});

test("Snapshot: Select control renders correctly with preselection of WA", () => {
  const tree = renderer.create(<FormPreselected />).toJSON();
  expect(tree).toMatchSnapshot();
});

/**
 * Interactive - Enzyme tests
 */
describe("SelectControl", function() {
  const expectedFieldsState = {
    state: {
      value: "",
      isValid: false,
      secondInteraction: false,
      errorMessage: "",
      validators: [validateRequired],
      optional: false
    }
  };
  const sharedFieldOptions = [
    {
      valueKey: "",
      valueText: "â–¾ Select a state"
    },
    {
      valueKey: "AL",
      valueText: "Alabama"
    },
    {
      valueKey: "AK",
      valueText: "Alaska"
    }
  ];

  it("should save initial field state", () => {
    const wrapper = mount(<Form />);
    const actual = wrapper.state("fields");
    expect(actual).toEqual(expectedFieldsState);
  });

  it("should take preselected fieldValue", function() {
    const wrapper = mount(<FormPreselected />);
    const selectedOption = wrapper.instance().getFieldState("state").value;
    expect(selectedOption).toBe("WA");
  });

  it("getFieldState should return fieldState", () => {
    const wrapper = mount(<Form />);
    const expected = expectedFieldsState.state;
    const actual = wrapper.instance().getFieldState("state");
    expect(actual).toEqual(expected);
  });

  it("should call handleValueChange if field has changed", function() {
    sinon.spy(Form.prototype, "handleValueChange");
    const wrapper = mount(<Form />);
    wrapper
      .find("select")
      .simulate("change", { target: { name: "state", value: "OR" } });
    expect(Form.prototype.handleValueChange.calledOnce).toEqual(true);
  });

  it("should call handleBlur if field is blurred", function() {
    sinon.spy(Form.prototype, "handleBlur");
    const wrapper = mount(<Form />);
    wrapper.find("select").simulate("blur");
    expect(Form.prototype.handleBlur.calledOnce).toEqual(true);
  });

  it("should call handleFocus if input is focus", function() {
    const handleFocus =  sinon.spy();

    const USStateFieldOptions = [
      {
        valueKey: "AL",
        valueText: "Alabama"
      },
      {
        valueKey: "AK",
        valueText: "Alaska"
      }];
 
    const SelectForm = props => {
      return (
        <form>
          <SelectControl
            fieldName="state"
            fieldOptions={USStateFieldOptions}
            fieldTitle="State"
            fieldClasses="wvus-field-33"
            formMethods={props.formMethods}
            handleFocus={handleFocus}
            validators={[validateRequired]}
          />
        </form>
      );
    };

    const Form = WVUSForm(SelectForm);
    
    const wrapper = mount(<Form />);
   // console.log(wrapper);
    wrapper.find("select").simulate("focus");
 
    expect(handleFocus.calledOnce).toEqual(true);
  });

  it("should show required star if required field", function() {
    const props = { formMethods: formMethodsUnTouched };
    const wrapper = mount(<SelectForm {...props} />);
    expect(wrapper.contains(<sup>*</sup>)).toBe(true);
  });

  it("should NOT show required star if optional field", function() {
    const props = { formMethods: formMethodsOptional };
    const wrapper = mount(<SelectForm {...props} />);
    expect(wrapper.contains(<sup>*</sup>)).toBe(false);
  });

  it("should render all field options passed via prop", function() {
    const props = {
      formMethods: formMethodsUnTouched,
      fieldOptions: sharedFieldOptions
    };
    const wrapper = mount(<SelectFormNoConfig {...props} />);
    expect(wrapper.find("option")).toHaveLength(sharedFieldOptions.length);
  });

  it("should have correct classes if optional field", function() {
    const props = {
      formMethods: formMethodsOptional,
      fieldOptions: sharedFieldOptions
    };
    const wrapper = mount(<SelectFormNoConfig {...props} />);
    expect(wrapper.find(".has-error")).toHaveLength(0);
    expect(wrapper.find(".has-success")).toHaveLength(0);
  });

  it("should have correct classes if has success", function() {
    const props = {
      formMethods: formMethodsWithSuccess,
      fieldOptions: sharedFieldOptions
    };
    const wrapper = mount(<SelectFormNoConfig {...props} />);
    expect(wrapper.find(".has-success")).toHaveLength(1);
    expect(wrapper.find(".has-error")).toHaveLength(0);
  });

  it("should have correct classes if has error", function() {
    const props = {
      formMethods: formMethodsWithError,
      fieldOptions: sharedFieldOptions
    };
    const wrapper = mount(<SelectFormNoConfig {...props} />);
    expect(wrapper.find(".has-error")).toHaveLength(1);
    expect(wrapper.find(".has-success")).toHaveLength(0);
  });

  it("should show error Message if has error", function() {
    const props = {
      formMethods: formMethodsWithError,
      fieldOptions: sharedFieldOptions
    };
    const wrapper = mount(<SelectFormNoConfig {...props} />);
    const errorMessage = props.formMethods.getFieldState().errorMessage;
    expect(
      wrapper.contains(
        <Message showError={true} showSuccess={false} message={errorMessage} />
      )
    ).toBe(true);
  });

  it("should show success Message if has success", function() {
    const props = {
      formMethods: formMethodsWithSuccess,
      fieldOptions: sharedFieldOptions
    };
    const wrapper = mount(<SelectFormNoConfig {...props} />);
    expect(
      wrapper.contains(<Message showError={false} showSuccess={true} />)
    ).toBe(true);
  });

  it("should be capable of having custom label and input classes", function() {
    const wrapper = mount(<Form />);
    expect(wrapper.find(".wvus-field-state label").hasClass("state-label")).toBe(true);
    expect(wrapper.find(".wvus-field-state select").hasClass("state-input")).toBe(true);
  });
});
