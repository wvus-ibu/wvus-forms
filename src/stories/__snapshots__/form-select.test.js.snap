// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot:  Select control renders correctly with no selection 1`] = `
<div
  className=""
>
  <form>
    <div
      className="wvus-field-state wvus-field-33 form-group has-icon "
    >
      <label
        className="state-label"
        htmlFor="state"
      >
        State
         
        <sup>
          *
        </sup>
      </label>
      <div
        className="form-control-wrapper"
      >
        <select
          className="form-control select-control state-input"
          id="state"
          name="state"
          onBlur={[Function]}
          onChange={[Function]}
          value=""
        >
          <option
            value=""
          >
            ▾ Select a state
          </option>
          <option
            value="AL"
          >
            Alabama
          </option>
          <option
            value="AK"
          >
            Alaska
          </option>
          <option
            value="AS"
          >
            American Samoa
          </option>
          <option
            value="AZ"
          >
            Arizona
          </option>
          <option
            value="AR"
          >
            Arkansas
          </option>
          <option
            value="AF"
          >
            Armed Forces Africa
          </option>
          <option
            value="AA"
          >
            Armed Forces Americas
          </option>
          <option
            value="AC"
          >
            Armed Forces Canada
          </option>
          <option
            value="AE"
          >
            Armed Forces Europe
          </option>
          <option
            value="AM"
          >
            Armed Forces Middle East
          </option>
          <option
            value="AP"
          >
            Armed Forces Pacific
          </option>
          <option
            value="CA"
          >
            California
          </option>
          <option
            value="CO"
          >
            Colorado
          </option>
          <option
            value="CT"
          >
            Connecticut
          </option>
          <option
            value="DE"
          >
            Delaware
          </option>
          <option
            value="DC"
          >
            District Of Columbia
          </option>
          <option
            value="FM"
          >
            Federated States Of Micronesia
          </option>
          <option
            value="FL"
          >
            Florida
          </option>
          <option
            value="GA"
          >
            Georgia
          </option>
          <option
            value="GU"
          >
            Guam
          </option>
          <option
            value="HI"
          >
            Hawaii
          </option>
          <option
            value="ID"
          >
            Idaho
          </option>
          <option
            value="IL"
          >
            Illinois
          </option>
          <option
            value="IN"
          >
            Indiana
          </option>
          <option
            value="IA"
          >
            Iowa
          </option>
          <option
            value="KS"
          >
            Kansas
          </option>
          <option
            value="KY"
          >
            Kentucky
          </option>
          <option
            value="LA"
          >
            Louisiana
          </option>
          <option
            value="ME"
          >
            Maine
          </option>
          <option
            value="MH"
          >
            Marshall Islands
          </option>
          <option
            value="MD"
          >
            Maryland
          </option>
          <option
            value="MA"
          >
            Massachusetts
          </option>
          <option
            value="MI"
          >
            Michigan
          </option>
          <option
            value="MN"
          >
            Minnesota
          </option>
          <option
            value="MS"
          >
            Mississippi
          </option>
          <option
            value="MO"
          >
            Missouri
          </option>
          <option
            value="MT"
          >
            Montana
          </option>
          <option
            value="NE"
          >
            Nebraska
          </option>
          <option
            value="NV"
          >
            Nevada
          </option>
          <option
            value="NH"
          >
            New Hampshire
          </option>
          <option
            value="NJ"
          >
            New Jersey
          </option>
          <option
            value="NM"
          >
            New Mexico
          </option>
          <option
            value="NY"
          >
            New York
          </option>
          <option
            value="NC"
          >
            North Carolina
          </option>
          <option
            value="ND"
          >
            North Dakota
          </option>
          <option
            value="MP"
          >
            Northern Mariana Islands
          </option>
          <option
            value="OH"
          >
            Ohio
          </option>
          <option
            value="OK"
          >
            Oklahoma
          </option>
          <option
            value="OR"
          >
            Oregon
          </option>
          <option
            value="PW"
          >
            Palau
          </option>
          <option
            value="PA"
          >
            Pennsylvania
          </option>
          <option
            value="PR"
          >
            Puerto Rico
          </option>
          <option
            value="RI"
          >
            Rhode Island
          </option>
          <option
            value="SC"
          >
            South Carolina
          </option>
          <option
            value="SD"
          >
            South Dakota
          </option>
          <option
            value="TN"
          >
            Tennessee
          </option>
          <option
            value="TX"
          >
            Texas
          </option>
          <option
            value="UT"
          >
            Utah
          </option>
          <option
            value="VT"
          >
            Vermont
          </option>
          <option
            value="VI"
          >
            Virgin Islands
          </option>
          <option
            value="VA"
          >
            Virginia
          </option>
          <option
            value="WA"
          >
            Washington
          </option>
          <option
            value="WV"
          >
            West Virginia
          </option>
          <option
            value="WI"
          >
            Wisconsin
          </option>
          <option
            value="WY"
          >
            Wyoming
          </option>
        </select>
        
        
      </div>
    </div>
  </form>
</div>
`;

exports[`Snapshot: Select control renders correctly with preselection of WA 1`] = `
<div
  className=""
>
  <form>
    <div
      className="wvus-field-state wvus-field-33 form-group has-icon has-success "
    >
      <label
        className=""
        htmlFor="state"
      >
        State
         
        <sup>
          *
        </sup>
      </label>
      <div
        className="form-control-wrapper"
      >
        <select
          className="form-control select-control "
          id="state"
          name="state"
          onBlur={[Function]}
          onChange={[Function]}
          value="WA"
        >
          <option
            value=""
          >
            ▾ Select a state
          </option>
          <option
            value="AL"
          >
            Alabama
          </option>
          <option
            value="AK"
          >
            Alaska
          </option>
          <option
            value="AS"
          >
            American Samoa
          </option>
          <option
            value="AZ"
          >
            Arizona
          </option>
          <option
            value="AR"
          >
            Arkansas
          </option>
          <option
            value="AF"
          >
            Armed Forces Africa
          </option>
          <option
            value="AA"
          >
            Armed Forces Americas
          </option>
          <option
            value="AC"
          >
            Armed Forces Canada
          </option>
          <option
            value="AE"
          >
            Armed Forces Europe
          </option>
          <option
            value="AM"
          >
            Armed Forces Middle East
          </option>
          <option
            value="AP"
          >
            Armed Forces Pacific
          </option>
          <option
            value="CA"
          >
            California
          </option>
          <option
            value="CO"
          >
            Colorado
          </option>
          <option
            value="CT"
          >
            Connecticut
          </option>
          <option
            value="DE"
          >
            Delaware
          </option>
          <option
            value="DC"
          >
            District Of Columbia
          </option>
          <option
            value="FM"
          >
            Federated States Of Micronesia
          </option>
          <option
            value="FL"
          >
            Florida
          </option>
          <option
            value="GA"
          >
            Georgia
          </option>
          <option
            value="GU"
          >
            Guam
          </option>
          <option
            value="HI"
          >
            Hawaii
          </option>
          <option
            value="ID"
          >
            Idaho
          </option>
          <option
            value="IL"
          >
            Illinois
          </option>
          <option
            value="IN"
          >
            Indiana
          </option>
          <option
            value="IA"
          >
            Iowa
          </option>
          <option
            value="KS"
          >
            Kansas
          </option>
          <option
            value="KY"
          >
            Kentucky
          </option>
          <option
            value="LA"
          >
            Louisiana
          </option>
          <option
            value="ME"
          >
            Maine
          </option>
          <option
            value="MH"
          >
            Marshall Islands
          </option>
          <option
            value="MD"
          >
            Maryland
          </option>
          <option
            value="MA"
          >
            Massachusetts
          </option>
          <option
            value="MI"
          >
            Michigan
          </option>
          <option
            value="MN"
          >
            Minnesota
          </option>
          <option
            value="MS"
          >
            Mississippi
          </option>
          <option
            value="MO"
          >
            Missouri
          </option>
          <option
            value="MT"
          >
            Montana
          </option>
          <option
            value="NE"
          >
            Nebraska
          </option>
          <option
            value="NV"
          >
            Nevada
          </option>
          <option
            value="NH"
          >
            New Hampshire
          </option>
          <option
            value="NJ"
          >
            New Jersey
          </option>
          <option
            value="NM"
          >
            New Mexico
          </option>
          <option
            value="NY"
          >
            New York
          </option>
          <option
            value="NC"
          >
            North Carolina
          </option>
          <option
            value="ND"
          >
            North Dakota
          </option>
          <option
            value="MP"
          >
            Northern Mariana Islands
          </option>
          <option
            value="OH"
          >
            Ohio
          </option>
          <option
            value="OK"
          >
            Oklahoma
          </option>
          <option
            value="OR"
          >
            Oregon
          </option>
          <option
            value="PW"
          >
            Palau
          </option>
          <option
            value="PA"
          >
            Pennsylvania
          </option>
          <option
            value="PR"
          >
            Puerto Rico
          </option>
          <option
            value="RI"
          >
            Rhode Island
          </option>
          <option
            value="SC"
          >
            South Carolina
          </option>
          <option
            value="SD"
          >
            South Dakota
          </option>
          <option
            value="TN"
          >
            Tennessee
          </option>
          <option
            value="TX"
          >
            Texas
          </option>
          <option
            value="UT"
          >
            Utah
          </option>
          <option
            value="VT"
          >
            Vermont
          </option>
          <option
            value="VI"
          >
            Virgin Islands
          </option>
          <option
            value="VA"
          >
            Virginia
          </option>
          <option
            value="WA"
          >
            Washington
          </option>
          <option
            value="WV"
          >
            West Virginia
          </option>
          <option
            value="WI"
          >
            Wisconsin
          </option>
          <option
            value="WY"
          >
            Wyoming
          </option>
        </select>
        
        <span>
          <i
            className="fa fa-check-circle"
          />
          <span
            className="message"
            data-field-validation-message={true}
          />
        </span>
      </div>
    </div>
  </form>
</div>
`;
